qcodes/instrument/channel.py

```diff
@@ -208,25 +214,45 @@ def __init__(self, parent: InstrumentBase,
                 raise TypeError("All items in this channel list must be of "
                                 "type {}.".format(chan_type.__name__))
 
+    def _shift_index(self, i: Optional[int]) -> Optional[int]:
+        """
+        Shifts an index by -index_origin unless it is None or negative.
+
+        Args:
+            i: int or None
+
+        Raises:
+            IndexError: If 0 <= i < index_origin
+        """
+        if i is None or i < 0:
+            return i
```

Comment:
So, per my understanding, the negative indexing (where ``-1`` means last, ``-2`` means 2nd to last, etc) should still "just" work regardless of the index origin, right? That's fine, BUT - why would one ever try to index with ``None``? I'd suggest that it should be an error; in all the methods where the _shift_index is used the argument ``i`` is always an ``int`` and never ``None`` (unless i missed something?). Hence I suggest the following implementation, and let's change the signature from ``_shift_index(self, i: Optional[int]) -> Optional[int]`` to ``_shift_index(self, i: int) -> int``:
```suggestion
        if i is None:
            raise ValueError("Cannot index with None")
        if i < 0:
            return i
```

Reply:
`_shift_index(i)` is called with `i = None` for example at line 246 when a ChannelList is sliced like `channel_list[:3]`, because in this case `channel_list.__getitem__(slice(None, 3, None))` is actually called.
A `None` means `0` when used as the start value and means `len(channel_list)` when used as the end value.
You are right that `i = None` should raise an error when indexing with an int or a tuple.
In the current version, the error is raised when indexing `self._channels`, after `_shift_index(i)` is called.
Do you think it is better to raise the error earlier?

Reply:
> A None means 0 when used as the start value and means len(channel_list) when used as the end value.

aa, i see, thanks for explaining!

> In the current version, the error is raised when indexing self._channels, after _shift_index(i) is called. Do you think it is better to raise the error earlier?

that's probably OK. But it puts more pressure on the implementation of _shift_index to propagate the ``None`` value through and don't try to be smart :) 

Let's do the following - let's keep the current implementation AND add a comment in the code of _shift_index explaining why ``None`` value is just being passed thorugh. Would you agree?

Reply:
I agree, the intention was not at all obvious. Give me a few minutes to come up with a good comment...